updates (#374)
allenai/OLMo-core: docs/source/guides/all_in_one_for_researchers.md

   For example, you can add the option `--data_loader.prefetch_factor=4` to update the `prefetch_factor` field within the `data_loader` part of the config.

   ```{tip}
   The script includes a dry-run mode that you can use to validate that your overrides are applied correctly.
   Just add the overrides option `--dry-run` and the script will print out the config and exit without actually training.
   If you pass an invalid overrides option (wrong type or invalid field name) the script will raise an exception.
   ```

### Launching the run

Now that we know how to change settings on the fly we're ready to launch the run.
And in order to get it running as fast as possible we're going to turn off a few features from the command-line that we'd normally want on, such as checkpointing and in-loop evals.
We're also going to tell the trainer to stop at step 100.
So the overrides we'll pass in are:

- `--trainer.callbacks.lm_evaluator.enabled=false` to disable the in-loop perplexity evaluator.
- `--trainer.callbacks.downstream_evaluator.enabled=false` to disable the in-loop downstream task evaluator.
- `--trainer.no_checkpoints` to disable checkpointing.

> - `--trainer.hard_stop='{value: 100, unit: steps}'` to stop at step 100.

Reviewer: huh, TIL

And see the [scaling](#scaling) section below for more ideas.

### Other bugs

For other bugs unrelated to CUDA, it's always a good idea to try to isolate the code that causes the issue.
If you can reproduce the issue in a small standalone script, it will be much easier to debug and fix.
And if that bug can be reproduced from a single process you could run it with a debugging like [pdb](https://docs.python.org/3/library/pdb.html).

Issues that only manifest in a distributed setting can be harder to debug.
Consider writing a distributed unit test with the {func}`~olmo_core.testing.run_distributed_test` helper function.

## Scaling

Scaling transformers is a complex topic.
A complete guide on the matter is well beyond the scope of this document, so this section will be focused on settings that can be applied to the {class}`~olmo_core.train.train_module.TransformerTrainModule` specifically, as we anticipate most readers will be using that.

But for more in-depth information on the topic we recommend checking out the [Scaling Book](https://jax-ml.github.io/scaling-book).

> As of writing this, the `TransformerTrainModule` supports 3 dimensions of parallelism: data parallelism through FSDP or DDP, tensor parallelism (TP), and context parallelism (CP).

Reviewer: ```suggestion
At the time of writing, the `TransformerTrainModule` supports 3 dimensions of parallelism: data parallelism through FSDP or DDP, tensor parallelism (TP), and context parallelism (CP).
```
nit

Also we dont support PP or EP? Or just not recommended until Tianhua's changes are in?
